onnx/common/path.h

```diff
@@ -8,17 +8,58 @@
 #pragma once
 
 #include <string>
+#ifdef _WIN32
+#include <windows.h>
+#endif
 
 namespace ONNX_NAMESPACE {
 
 #ifdef _WIN32
 const std::string k_preferred_path_separator = "\\";
+const std::wstring w_k_preferred_path_separator = L"\\";
 #else // POSIX
 const std::string k_preferred_path_separator = "/";
 #endif
 
-std::string path_join(const std::string& origin, const std::string& append);
-void normalize_separator(std::string& path);
-std::string clean_relative_path(const std::string& path);
+template <typename STRING>
+STRING path_join(const STRING& origin, const STRING& append);
+
+template <typename CHAR>
+bool is_path_separator(CHAR c) {
+  // Windows accept / as path separator.
+  if (k_preferred_path_separator == "\\") {
+    return c == '\\' || c == '/';
+  }
+
+  return c == k_preferred_path_separator[0];
+}
+
+template <typename STRING>
+void normalize_separator(STRING& path) {
+  char preferred_sep = k_preferred_path_separator[0];
+  if (preferred_sep == '/') {
+    // Do nothing on linux.
+    return;
+  }
+
+  for (size_t i = 0; i < path.size(); i++) {
+    if (is_path_separator(path[i]) && path[i] != preferred_sep) {
+      path[i] = preferred_sep;
+    }
+  }
+}
+template <typename STRING>
+STRING clean_relative_path(const STRING& path);
+std::string clean_relative_path(const char* path);
+
+#ifdef _WIN32
+std::wstring clean_relative_path(const wchar_t* path);
+inline std::wstring utf8str_to_wstring(const std::string& utf8str) {
+  int size_required = MultiByteToWideChar(CP_UTF8, 0, utf8str.c_str(), (int)utf8str.size(), NULL, 0);
+  std::wstring ws_str(size_required, 0);
```

Comment:
Please use L'\0'

Reply:
utf8str.size() doesn't include the nul-terminator of the string. 
So does `size_required`.
When you construct ws_str with length=size_required, will the underlaying implementation automatically add a '\0' at the end? I'm not sure.

Reply:
Good point. I checked the document and it does not explicitly say we need to handle end of string on our own so I tried to rum some experiments as you suggested. I initialize some Chinese+English characters as std::wstring and then use `std::codecvt_utf8` to convert them from std::wstring into std::string. Next, I used the `utf8str_to_wstring` here to convert these std::string back to std::wstring. In result, the twice converted std::wstring is identical as the one before conversion. Therefore, I think the original method should be good. Please let me know if you still have concern. Thanks!

Reply:
LGTM.